### PR TITLE
Extra process guards

### PR DESCRIPTION
**Pull request by heckj**
Following on from earlier PR (https://github.com/apple/swift-package-manager/pull/921) and bug SR-3275 - adding in the mechanisms to support the extra guard logic for validating that an executable file (i.e. git, etc) exists, and making the results when it doesn't exist more consistent across all the various underlying toolchains.

Here for visibility and feedback...

Next steps:
- [x] rebase after https://github.com/apple/swift-package-manager/pull/1002 & https://github.com/apple/swift-package-manager/pull/1003 merged
- [x] add in string output handling for new errors enums in Commands/Errors.swift
- [x] add in relevant tests to validate

### CODE REVIEW

File: Sources/Basic/FileSystem.swift [hunk 1]
```diff
@@ -101,6 +101,9 @@ public protocol FileSystem {
     
     /// Check whether the given path is accessible and a file.
     func isFile(_ path: AbsolutePath) -> Bool
+    
+    /// Check to determine whether the given path is an accessible and executable file
+    func isExecutableFile(_ path: AbsolutePath) -> Bool
```
Comment (aciidb0mb3r) [inline Sources/Basic/FileSystem.swift]: Please create a separate PR for this change. We try to keep all orthogonal changes (especially API addition in Basic) in different PRs.

File: Sources/Basic/FileSystem.swift [hunk 2]
```diff
@@ -155,6 +158,18 @@ public extension FileSystem {
 
 /// Concrete FileSystem implementation which communicates with the local file system.
 private class LocalFileSystem: FileSystem {
+    
+    /// Check to determine whether the given path is an accessible and executable file
```
Comment (aciidb0mb3r) [inline Sources/Basic/FileSystem.swift]: Don't require the doc comment here.

File: Sources/Basic/FileSystem.swift [hunk 3]
```diff
@@ -101,6 +101,9 @@ public protocol FileSystem {
     
     /// Check whether the given path is accessible and a file.
     func isFile(_ path: AbsolutePath) -> Bool
+    
+    /// Check to determine whether the given path is an accessible and executable file
```
Comment (aciidb0mb3r) [inline Sources/Basic/FileSystem.swift]: Please end all comments with punctuation.

File: Sources/SourceControl/GitRepository.swift [hunk 4]
```diff
@@ -600,6 +600,17 @@ private class GitFileSystemView: FileSystem {
...
     
+    func isExecutableFile(_ path: AbsolutePath) -> Bool {
+        do {
+            if let entry = try getEntry(path), entry.type == .executableBlob {
+                return true
+            }
+            return false
+        } catch {
+            return false
+        }
```
Comment (aciidb0mb3r) [inline Sources/SourceControl/GitRepository.swift]: I think this can be simplified to:
```swift
if let entry = try? getEntry(path), entry.type == .executableBlob {
    return true
}
return false
```

File: Sources/Utility/Process.swift [hunk 5]
```diff
@@ -373,6 +373,98 @@ extension Process {
...
+            // else throw ProgramSearchError.couldNotFindProgram
+//}
+
+/// Create a list of AbsolutePath search paths from a string, such as the PATH environment variable.
+///
+/// - Parameters:
+///   - pathString: The path string to parse.
+///   - currentWorkingDirectory: The current working directory, the relative paths will be converted to absolute paths based on this path.
+/// - Returns: List of search paths.
+public func getEnvSearchPaths(
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Please propose this move in a separate if this PR is going to have additional commits

File: Tests/BasicTests/FileSystemTests.swift [hunk 6]
```diff
@@ -42,10 +43,23 @@ class FileSystemTests: XCTestCase {
...
         XCTAssertFalse(fs.isDirectory(sym))
 
+        // isExecutableFile
+        let executable = tempDir.path.appending(component: "exec-foo")
+        let stream = BufferedOutputByteStream()
+        stream <<< "#!/bin/sh" <<< "\n"
+        stream <<< "set -e" <<< "\n"
+        stream <<< "exit" <<< "\n"
+        try! localFileSystem.writeFileContents(executable, bytes: stream.bytes)
+        _ = try! Process.popen(args: "chmod", "+x", executable.asString)
```
Comment (aciidb0mb3r) [inline Tests/BasicTests/FileSystemTests.swift]: use `Process.checkNonZeroExit()` because this will not fail if chmod exits with non zero code.

File: Sources/SourceControl/GitRepository.swift (hunk 4 above)
Comment (heckj) [inline Sources/SourceControl/GitRepository.swift]: the 3 preceding bits of logic all follow that same pattern - update them all to keep it consistent? (`isDirectory`, `isFile`, and `isSymlink`) - use the `try?` on all of them instead of the longer `do {} catch {}` block structure?

File: Sources/Basic/FileSystem.swift (hunk 1 above)
Comment (heckj) [inline Sources/Basic/FileSystem.swift]: will do...

File: Sources/Basic/FileSystem.swift (hunk 3 above)
Comment (heckj) [inline Sources/Basic/FileSystem.swift]: thanks, will fix

File: Sources/Basic/FileSystem.swift (hunk 2 above)
Comment (heckj) [inline Sources/Basic/FileSystem.swift]: will clear

File: Sources/Utility/Process.swift (hunk 5 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: another PR for the migration of the search paths bits from UserControls -> Process coming up

File: Tests/BasicTests/FileSystemTests.swift (hunk 6 above)
Comment (heckj) [inline Tests/BasicTests/FileSystemTests.swift]: Ah! Thank you!

File: Sources/Utility/Process.swift (hunk 5 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: moved environment search path logic (and related tests) to another PR: https://github.com/apple/swift-package-manager/pull/996

File: Sources/Basic/FileSystem.swift (hunk 1 above)
Comment (heckj) [inline Sources/Basic/FileSystem.swift]: migrated the Filesystem protocol extension, implementation and related tests to another PR: https://github.com/apple/swift-package-manager/pull/997

**Comment by heckj** [comment]
@aciidb0mb3r PR updated after the various merges - still need to write tests, but wanted to get your eyes on this first. Was tough choosing where to put in a relevant check to validate `git` that wasn't in 1000 places, but also wasn't checked when not needed. I aimed for a putting it in `getActiveWorkspace` (which looked like it was always in the right flow for the places where this check would be valuable) with the idea that any of the tool's running `runImpl` would leverage that method and get the check, and I could throw an error from there, and update the Error handling output in Commands/Error.swift to print appropriately. This sound like an acceptable plan of attack, and stubbed/initial implementation?

File: Sources/Commands/SwiftTool.swift [hunk 7]
```diff
@@ -203,6 +204,17 @@ public class SwiftTool<Options: ToolOptions> {
...
+        let envSearchPaths = Utility.getEnvSearchPaths(
+            pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
+
+        guard let gitLocation = Utility.lookupExecutablePath(filename: Git.tool, searchPaths: envSearchPaths) else {
+            throw SwiftToolError.missingGitDependency
+            // NOTE(heckj): differentiate error reporting between not finding it and it not being executable?
+        }
+        guard localFileSystem.isExecutableFile(gitLocation) else {
+            throw SwiftToolError.missingGitDependency
+        }
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftTool.swift]: We should place this check in Process class (just before calling posix_spawn) so every subprocess we spawn is checked and not make it specialised for git.

File: Sources/Commands/UserToolchain.swift [hunk 8]
```diff
@@ -57,21 +57,19 @@ public struct UserToolchain: Toolchain, ManifestResourceProvider {
...
         // First look in env and then in bin dir.
-        swiftCompiler = lookup(env: "SWIFT_EXEC") ?? binDir.appending(component: "swiftc")
+        swiftCompiler = lookup(fromEnv: "SWIFT_EXEC") ?? binDir.appending(component: "swiftc")
         
-        // Check that it's valid in the file system.
-        // FIXME: We should also check that it resolves to an executable file
-        //        (it could be a symlink to such as file).
-        guard localFileSystem.exists(swiftCompiler) else {
+        // Check that it's valid in the file system and executable.
+        guard localFileSystem.isExecutableFile(swiftCompiler) else {
```
Comment (aciidb0mb3r) [inline Sources/Commands/UserToolchain.swift]: Please propose completely unrelated changes (such as this and lookup rename) in a different PR.

File: Sources/Commands/SwiftTool.swift (hunk 7 above)
Comment (heckj) [inline Sources/Commands/SwiftTool.swift]: I think this may be a preferrable solution, not because it's specific to git, but because it only happens once and the beginning of any invocation, regardless of how many times "git" is invoked - which in a larger project could be quite a number of times. Vapor lists 6 dependencies for example - 5 times more checks than is strictly needed to validate Git is around and in the path as an executable. 

If you insist that we should do this check as a part of Process, I'll make the changes and updates, but I'd like to ask that you reconsider.

File: Sources/Commands/UserToolchain.swift (hunk 8 above)
Comment (heckj) [inline Sources/Commands/UserToolchain.swift]: They didn't seem unrelated to me (the process of adding extra guards/checks to the code), but happy to move to another PR

File: Sources/Commands/SwiftTool.swift (hunk 7 above)
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftTool.swift]: Thats a good point but I still think this belongs in Process class, we can cache the results so we don't need to check in subsequent calls. I'll leave it upto you if you want to implement he cache right now (it can just be a simple dictionary).

File: Sources/Commands/UserToolchain.swift (hunk 8 above)
Comment (aciidb0mb3r) [inline Sources/Commands/UserToolchain.swift]: Ah I see, thats a fair categorization. I was thinking about this PR as a feature addition to Process class (which I still do, see previous comment).

Comment (heckj) [inline Sources/Commands/UserToolchain.swift]: No worries, easy to shift this to another PR

File: Sources/Commands/SwiftTool.swift (hunk 7 above)
Comment (heckj) [inline Sources/Commands/SwiftTool.swift]: Gotcha, can and will do, thanks! I'll shift it over and set up the relevant testing.

File: Sources/Commands/UserToolchain.swift (hunk 8 above)
Comment (heckj) [inline Sources/Commands/UserToolchain.swift]: moved these updates to https://github.com/apple/swift-package-manager/pull/1003

**Comment by heckj** [comment]
@aciidb0mb3r I think it's ready for a review - I can rebase to a single commit before merging, but wanted to get feedback from here. 

I leveraged the other PR's adding in the various guards, and this wraps around Process to check for errors using that simple cache mechanism. I've also added some messaging and handling into Commands/Error.swift to cover the execution failure within Process

File: Sources/Utility/Process.swift [hunk 9]
```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
...
+
+        // check to see if the argument is already listed in the vaildated cache
+        guard !Process.validatedExecutables.contains(program) else {
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
+        guard let executableLocation = Utility.lookupExecutablePath(filename: program, searchPaths: envSearchPaths) else {
+            throw Process.Error.missingProgram(program: program)
+        }
+        guard localFileSystem.isExecutableFile(executableLocation) else {
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: I think lookupExecutable ensure this.

File: Sources/Utility/Process.swift [hunk 10]
```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
...
+            throw Process.Error.programNotExecutable(program: program)
+        }
+        // add the program name to the validated cache
+        Process.validatedExecutables.insert(program)
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Hmm this cache needs to be thread safe. We should also store the result, right now we will recheck for the executables we did not find before.

File: Sources/Utility/Process.swift [hunk 11]
```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
         self.redirectOutput = redirectOutput 
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    internal func validateArgumentIsExecutable(_ program: String) throws {
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Prefer not to explicitly write internal.

File: Sources/Utility/Process.swift [hunk 12]
```diff
@@ -151,6 +155,9 @@ public final class Process: ObjectIdentifierProtocol {
     /// Queue to protect concurrent reads.
     private let serialQueue = DispatchQueue(label: "org.swift.swiftpm.process")
 
+    /// cache of validated executables arguments
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Please use proper capitalization, punctuation in all comments. http://llvm.org/docs/CodingStandards.html#commenting

File: Tests/UtilityTests/ProcessTests.swift [hunk 13]
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
         }
     }
 
+    func testProcessExecutableLookup() throws {
```
Comment (aciidb0mb3r) [inline Tests/UtilityTests/ProcessTests.swift]: Please add this in allTests variable otherwise it won't run on linux.

File: Tests/UtilityTests/ProcessTests.swift [hunk 14]
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
...
+        mktmpdir { path in
+            // should always work
+            try! Process().validateArgumentIsExecutable("ls")
+
+            // should report missingProgram to non-existant program
+            do {
+                try Process().validateArgumentIsExecutable("nonExistantProgram")
+                XCTFail("Should have failed to validate nonExistantProgram")
+            } catch Process.Error.missingProgram (let program) {
+                XCTAssertTrue(program.contains("nonExistantProgram"))
```
Comment (aciidb0mb3r) [inline Tests/UtilityTests/ProcessTests.swift]: this can check for equality, right?

File: Sources/Utility/Process.swift [hunk 15]
```diff
@@ -109,6 +110,9 @@ public final class Process: ObjectIdentifierProtocol {
     public enum Error: Swift.Error {
         /// Missing arguments to invoke a process
         case missingArguments
+        case missingProgram(program: String)
+        case programNotExecutable(program: String)
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Please add doc comments to these

Comment (heckj) [inline Sources/Utility/Process.swift]: will do

File: Sources/Utility/Process.swift (hunk 12 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: Ah, thank you for the link on the conventions, will update

File: Sources/Utility/Process.swift (hunk 11 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: What would be preferred? I made it `internal` so that I could test it - when I started, I was making it `private`, but then even @testable import Utility couldn't resolve it to test it (a bit surprising to me philosophically - I'm used to testing private methods in other languages). I could just make it public, but it didn't seem like it necessarily fit in that scenario. Very open to feedback though - what do you think would make sense here?

File: Sources/Utility/Process.swift (hunk 9 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: if you mean `lookupExecutablePath`, then unfortunately no - it doesn't. It's poorly named, but I've left the name in place (perhaps I should update it?) - that method only iterates through the search paths to find the executable, it doesn't actually *verify* it's executable as well.

I could combine these together and update that function, and return a single error (missing or not executable), but with them separate I can be more specific about the error. I honestly wasn't sure which was better, so i aimed for more information to debug/understand the problem if one happens.

File: Sources/Utility/Process.swift (hunk 10 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: I'm not entirely sure what you mean by "store the result", given that line 190 is appending the name of the program searched for, and found, in the Set. Do you mean store the result of where we found the executable? Even though we don't pass that back or use it? Or is that something we should be doing?

I thought about the thread-safe, and worst case that I could spot was that extra checking would happen, and the Set might be updated with an additional redundant value (and subsequently ignore it I presume). I'm feeling a little bait-and-switched with this implementation though, because my original proposal wouldn't have had to concern itself so much with thread safety since it was very early and simply in the execution path and before the system would start to attempt to do "interesting things" in parallel for efficiency. So - that said - I'm also not immediately clueful about *how* to make this thread safe, could you perhaps suggest a specific path to enabling that? Are you thinking that we should wrap the shared state (the cache) in a lock/protection akin to what is described at https://www.bignerdranch.com/blog/locking-in-swift-helping-protect-me-from-myself/ or some of the answers at http://stackoverflow.com/questions/28191079/create-thread-safe-array-in-swift?

File: Tests/UtilityTests/ProcessTests.swift (hunk 13 above)
Comment (heckj) [inline Tests/UtilityTests/ProcessTests.swift]: will do, sorry - missed that

File: Tests/UtilityTests/ProcessTests.swift (hunk 14 above)
Comment (heckj) [inline Tests/UtilityTests/ProcessTests.swift]: duh, yes - of course it can. Will update

File: Sources/Utility/Process.swift (hunk 11 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: oh I meant that the default scope is internal and we don't explicitly write internal in swiftpm's codebase.
For e.g., these two structs have similar scope:
```swift
struct Foo {}
```
```swift
internal struct Foo{}
```

File: Sources/Utility/Process.swift (hunk 9 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: In general, feel free to rename a methods to clarify their intents.

For this specific case, I actually don't want a specific error because there could be a lot of reason for failure... it could be a dir, we might not have permission, the mode of file may be invalid. Its probably better to just say "we couldn't find the executable you wanted" even tho thats a bit vague. What do you think?

File: Sources/Utility/Process.swift (hunk 10 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: By store the result I meant, store the result of the lookup query. Since we don't need the actual path, we can store the result as a bool where true means we found it, false means we did not find it.

For now, lets leave the cache and it can be an add on later. I'll point you to examples in SwiftPM for thread safety.

File: Sources/Utility/Process.swift (hunk 11 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: /me facepalm - got it.

File: Sources/Utility/Process.swift (hunk 10 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: cool -I'd like that. Something I'm not yet really understanding down to how to implement it, but I'd like to.

File: Sources/Utility/Process.swift (hunk 9 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: Based on my own past experiences of not quite understanding why something failed, I was leaning towards being as absolutely explicit about the error as I could be, presuming that for a CLI, a more detailed error would be useful to someone getting their development or CI environment debugged, and in the case of a library that the additional error types would be more useful. 

That said, the only places currently using this are in UserToolchain and now here in process, so if the specificity doesn't seem really needed, I can collapse it down to a "missing or not executable" pretty easily and shorten even this checking code. I'll update `Utility/Misc.swift` to have `lookupExecutablePath` only return a value if it's both found & executable, which can reduce the number of subsequent guards here and simplify this a bit.

**Comment by heckj** [comment]
@aciidb0mb3r simplified and updated, including the lookupExecutable method in misc and corresponding tests. Also rebased the whole kit to a single commit. I haven't make this explicitly thread-safe, so no improvements on that front from earlier commits, but I think it's in good order as it stands.

File: Sources/Utility/Process.swift [hunk 16]
```diff
@@ -107,8 +108,10 @@ public final class Process: ObjectIdentifierProtocol {
 
     /// Errors when attempting to invoke a process
     public enum Error: Swift.Error {
-        /// Missing arguments to invoke a process
+        /// The Process is missing arguments to invoke a process.
         case missingArguments
+        /// The program requested to be executed can not be found on the existing search paths, or is not executable.
```
Comment (xwu) [inline Sources/Utility/Process.swift]: Drive-by nits: "cannot" is one word.

Comment (xwu) [inline Sources/Utility/Process.swift]: Drive-by nits: "cannot" is one word.

File: Sources/Utility/Process.swift [hunk 17]
```diff
@@ -151,6 +154,9 @@ public final class Process: ObjectIdentifierProtocol {
     /// Queue to protect concurrent reads.
     private let serialQueue = DispatchQueue(label: "org.swift.swiftpm.process")
 
+    /// Cache of validated executables arguments to be checked againstfrom `validateArgumentIsExecutable`.
```
Comment (xwu) [inline Sources/Utility/Process.swift]: "againstfrom"?

File: Sources/Utility/Process.swift [hunk 18]
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
         self.redirectOutput = redirectOutput 
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    func validateArgumentIsExecutable(_ program: String) throws {
+
+        // check to see if the argument is already listed in the vaildated cache
```
Comment (xwu) [inline Sources/Utility/Process.swift]: These comments should probably be correctly capitalized and punctuated.

File: Sources/Utility/Process.swift [hunk 19]
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
...
+        guard !Process.validatedExecutables.contains(program) else {
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
+        guard Utility.lookupExecutablePath(filename: program, searchPaths: envSearchPaths) != nil else {
+            throw Process.Error.missingExecutableProgram(program: program)
+        }
+        // add the program name to the validated cache since it has been found
```
Comment (xwu) [inline Sources/Utility/Process.swift]: Ditto.

File: Tests/UtilityTests/ProcessTests.swift [hunk 20]
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
         }
     }
 
+    func testProcessExecutableLookup() throws {
+        mktmpdir { path in
+            // should always work
+            try! Process().validateArgumentIsExecutable("ls")
+
+            // should report missingProgram to non-existant program
```
Comment (xwu) [inline Tests/UtilityTests/ProcessTests.swift]: Typo: "nonexistent"

File: Tests/UtilityTests/ProcessTests.swift [hunk 21]
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
...
+            } catch Process.Error.missingExecutableProgram (let program) {
+                XCTAssert(program == "nonExistantProgram")
+            }
+
+            // create a local, non-executable file to test
+            let tempExecutable = path.appending(component: "nonExecutableProgram")
+            try localFileSystem.writeFileContents(tempExecutable, bytes: "#!/bin/sh\nexit\n")
+
+            try withCustomEnv(["PATH": path.asString]) {
+                // should report nonExecutable
```
Comment (xwu) [inline Tests/UtilityTests/ProcessTests.swift]: typo: "non-executable"

File: Tests/UtilityTests/miscTests.swift [hunk 22]
```diff
@@ -63,9 +65,10 @@ class miscTests: XCTestCase {
             exec = lookupExecutablePath(filename: "clang", currentWorkingDirectory: path, searchPaths: pathEnv)
             XCTAssertEqual(exec, pathEnvClang)
             
-            // Create the binary relative to cwd.
+            // Create the binary relative to cwd and make it executable
```
Comment (xwu) [inline Tests/UtilityTests/miscTests.swift]: Punctuation.

**Comment by heckj** [comment]
thanks @xwu appreciate the call-outs - updated.

File: Sources/Utility/Process.swift [hunk 23]
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
         self.redirectOutput = redirectOutput 
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    func validateArgumentIsExecutable(_ program: String) throws {
+
+        // Check to see if the argument is already listed in the vaildated cache.
+        guard !Process.validatedExecutables.contains(program) else {
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: Please remove the cache because it is not thread safe. It is better to do more than be unsafe,  as it can cause subtle bugs and crashes.

File: Sources/Commands/Error.swift [hunk 24]
```diff
@@ -151,7 +151,10 @@ private func _handle(_ error: Any) {
         }
         stream <<< result.arguments.map{$0.shellEscaped()}.joined(separator: " ")
         print(error: stream.bytes.asString!)
-        
+
+    case Process.Error.missingExecutableProgram(let program):
+        print(error: "Unable to find an executable \(program).")
```
Comment (aciidb0mb3r) [inline Sources/Commands/Error.swift]: It's actually weird but we don't capitalise the errors (or end them with punctuation). I don't know the reason but it's consistent with compiler.

File: Sources/Utility/Process.swift [hunk 25]
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
...
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
```
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: actually getenv and currentworkingdir is also not thread safe, i'll comment here again on how to make this thread safe.

Comment (heckj) [inline Sources/Utility/Process.swift]: Cool, thanks

File: Sources/Utility/Process.swift (hunk 23 above)
Comment (heckj) [inline Sources/Utility/Process.swift]: Will do

File: Sources/Commands/Error.swift (hunk 24 above)
Comment (heckj) [inline Sources/Commands/Error.swift]: Huh, OK - will update

File: Sources/Utility/Process.swift (hunk 25 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: I haven't tried compiling this but we need to do something similar to this:
```swift
    /// Queue to protect reading/writing on executables map.
    private static let executablesQueue = DispatchQueue(
            label: "org.swift.swiftpm.process.findExecutable")

    /// Contains results of executable lookup.
    ///
    /// Key: Executable name or path.
    /// Value: If key was found in the search paths.
    static private var executablesMap: [String: Bool] = [:]

    /// Returns true if the given program is present in search path.
    ///
    /// The program can be executable name, relative path or absolute path.
    func findExecutable(_ program: String) -> Bool {
        return Process.executablesQueue.sync {
            // Check if we already have a value for the program.
            if let value = Process.executablesMap[program] {
                return value
            }
            // FIXME: This can be cached.
            let envSearchPaths = Utility.getEnvSearchPaths(
                pathString: getenv("PATH"),
                currentWorkingDirectory: currentWorkingDirectory
            )
            // Lookup the executable.
            let value = Utility.lookupExecutablePath(
                filename: program, searchPaths: envSearchPaths) != nil
            Process.executablesMap[program] = value
            return value
        }
    }

```

Comment (heckj) [inline Sources/Utility/Process.swift]: Thanks @aciidb0mb3r, I'll update to work out doing this - and see about adding the caching for the path lookups as well.

So the general mechanism here, if I'm groking it, is to create a DispatchQueue and set it up as a class property, so there's just one of these in the system, and then when we need to run something that leveraging things needing thread safety, to do that relevant work within a call to the DispatchQueue `sync` method, handing in an executable block that returns a value.

I started reading through the guides on DispatchQueues, and the code examples there were in C and talked to creating either Serial or Concurrent quyeyeqs. - I wasn't clear what the initializer in swift returned - does that by default create a serial one? I'll keep reading, but thought I'd ask if you any good pointers to more info here.

File: Sources/Commands/Error.swift (hunk 24 above)
Comment (heckj) [inline Sources/Commands/Error.swift]: looking at the rest of this file, it's pretty mixed - there's a fair bit of initial capitalization, although much fewer punctuation ending the errors. I removed the punctuation, but left the capitalization since it seemed to fit with more other lines.

File: Sources/Utility/Process.swift [hunk 26]
```diff
@@ -165,6 +178,35 @@ public final class Process: ObjectIdentifierProtocol {
...
+            // Lookup the executable.
+            let value = Utility.lookupExecutablePath(
+                filename: program, searchPaths: envSearchPaths) != nil
+            Process.validatedExecutablesMap[program] = value
+            return value
+        }
+    }
+
+    /// Validate that the program provided exists on the search path and is executable.
+    func validateArgumentIsExecutable(_ program: String) throws {
```
Comment (heckj) [inline Sources/Utility/Process.swift]: I wasn't sure how (or if) to mix throwing exceptions with methods that are invoking synchronous commands on a DispatchQueue, I'm presuming the guard would still be valuable, but is it worth a method here to make it explicitly testable? Or would it make more sense to shift this down to a guard statement just directly in the `launch` method below, and update the tests to validate the boolean results of the search extraneous of testing for throwing exceptions to be handled? Do you have a preference for how that's structured?

File: Sources/Utility/Process.swift (hunk 25 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: The type of variable (class or property) depends on the use case, since we require that this method should be protected across all the instance of Process, it should be a class property.

Yes, by default, a serial queue is created and then you can dispatch work items on those queues. I haven't read this one specifically but tutorials here are generally good: https://www.raywenderlich.com/148513/grand-central-dispatch-tutorial-swift-3-part-1

File: Sources/Utility/Process.swift (hunk 26 above)
Comment (aciidb0mb3r) [inline Sources/Utility/Process.swift]: We should move the guard to launch() and just write test for `findExecutable()` and `launch()`

**Comment by heckj** [comment]
I suspected that might be the path you suggested. I'll update

**Comment by heckj** [comment]
@aciidb0mb3r give it another look-see? Changes made, methods collapsed, and rebased onto latest master into a single commit.

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
@heckj Thanks for working on this!

**Comment by heckj** [comment]
Thank you for the guidance and pointers on threadsafety!
